Env vars that are numbers are read as strings, causing bugs in patient matching

2022-01-14 15:29:09 +0000] [5] [ERROR] Error handling request /match
Traceback (most recent call last):
File "/home/worker/venv/lib/python3.8/site-packages/gunicorn/workers/gthread.py", line 271, in handle
keepalive = self.handle_request(req, conn)
File "/home/worker/venv/lib/python3.8/site-packages/gunicorn/workers/gthread.py", line 323, in handle_request
respiter = self.wsgi(environ, resp.start_response)
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 2091, in __call__
return self.wsgi_app(environ, start_response)
File "/home/worker/venv/lib/python3.8/site-packages/werkzeug/middleware/proxy_fix.py", line 187, in __call__
return self.app(environ, start_response)
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 2076, in wsgi_app
response = self.handle_exception(e)
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
response = self.full_dispatch_request()
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 1518, in full_dispatch_request
rv = self.handle_user_exception(e)
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request
rv = self.dispatch_request()
File "/home/worker/venv/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request
return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
File "/home/worker/venv/lib/python3.8/site-packages/flask_negotiate.py", line 14, in wrapper
return fn(*args, **kwargs)
File "/home/worker/venv/lib/python3.8/site-packages/flask_negotiate.py", line 27, in wrapper
return fn(*args, **kwargs)
File "/home/worker/app/patientMatcher/server/views.py", line 283, in match_internal
match_obj = internal_matcher(
File "/home/worker/app/patientMatcher/match/handler.py", line 108, in internal_matcher
if score["patient"] >= score_threshold:
TypeError: '>=' not supported between instances of 'float' and 'str'
10.0.2.100 - - [14/Jan/2022:15:29:09 +0000] "POST /match HTTP/1.0" 500 0 "-" "-"
[2022-01-14 15:35:11 +0000] [5] [DEBUG] GET /

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+## [] -
+### Fixed
+- Parsing of numerical and boolean env vars when creating the app 
+
 ## [3.1] - 2022-01-13
 ### Changed
 - Speed-up CI tests by caching installation of libs and splitting tests into randomized groups using pytest-test-groups

--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -35,10 +35,10 @@ services:
       dockerfile: Dockerfile-server
     environment:
       MONGODB_HOST: mongodb
-      MONGODB_PORT: 27017
-      MAX_GT_SCORE: 0.75
-      MAX_PHENO_SCORE: 0.25
-      MAX_RESULTS: 10
+      MONGODB_PORT: '27017'
+      MAX_GT_SCORE: '0.75'
+      MAX_PHENO_SCORE: '0.25'
+      MAX_RESULTS: '10'
       DISCLAIMER: 'This is a test PatientMatcher server'
       ADMINS: '["test@mail.se"]'
       TESTING: 'True'

--- a/patientMatcher/instance/config.py
+++ b/patientMatcher/instance/config.py
@@ -1,10 +1,10 @@
 import os
 
 # Turns on debugging features in Flask
-DEBUG = os.getenv("DEBUG") or True
+DEBUG = os.getenv("DEBUG", "True") == "True" or True
 
 # Set to False is app if a production app
-TESTING = os.getenv("TESTING") or True
+TESTING = os.getenv("TESTING", "True") == "True" or True
 
 # secret key:
 SECRET_KEY = os.getenv("SECRET_KEY") or "MySuperSecretKey"
@@ -15,7 +15,7 @@ DB_NAME = os.getenv("MONGODB_DBNAME") or "pmatcher"
 DB_HOST = (
     os.getenv("MONGODB_HOST") or "127.0.0.1"
 )  # simply substitute with 'mongodb' if connecting to MongoDB running in a container
-DB_PORT = os.getenv("MONGODB_PORT") or 27017
+DB_PORT = int(os.getenv("MONGODB_PORT")) if os.getenv("MONGODB_PORT") else 27017
 
 # DB_URI = "mongodb://{}:{}@{}:{}/{}".format(DB_USERNAME, DB_PASSWORD, DB_HOST, DB_PORT, DB_NAME)
 DB_URI = os.getenv("MONGODB_DB_URI") or "mongodb://{}:{}/{}".format(
@@ -25,15 +25,15 @@ DB_URI = os.getenv("MONGODB_DB_URI") or "mongodb://{}:{}/{}".format(
 
 # Matching Algorithms scores
 # sum of MAX_GT_SCORE and MAX_PHENO_SCORE should be 1
-MAX_GT_SCORE = os.getenv("MAX_GT_SCORE") or 0.75
-MAX_PHENO_SCORE = os.getenv("MAX_PHENO_SCORE") or 0.25
+MAX_GT_SCORE = float(os.getenv("MAX_GT_SCORE")) if os.getenv("MAX_GT_SCORE") else 0.75
+MAX_PHENO_SCORE = float(os.getenv("MAX_PHENO_SCORE")) if os.getenv("MAX_PHENO_SCORE") else 0.25
 
 # Max results matches returned by server.
-MAX_RESULTS = os.getenv("MAX_RESULTS") or 5
+MAX_RESULTS = int(os.getenv("MAX_RESULTS")) if os.getenv("MAX_RESULTS") else 5
 
 # Set a minimum patient score threshold for returned results
 # Set this parameter to 0 to return all results with a score higher than 0
-SCORE_THRESHOLD = os.getenv("SCORE_THRESHOLD") or 0
+SCORE_THRESHOLD = float(os.getenv("SCORE_THRESHOLD")) if os.getenv("SCORE_THRESHOLD") else 0
 
 # Disclaimer. This text is returned along with match results or server metrics
 DISCLAIMER = (
@@ -44,9 +44,9 @@ DISCLAIMER = (
 # Email notification params.
 # Required only if you want to send match notifications to patients contacts
 MAIL_SERVER = os.getenv("MAIL_SERVER") or None  # "smtp.gmail.com"
-MAIL_PORT = os.getenv("MAIL_PORT") or None  # 587
-MAIL_USE_TLS = os.getenv("MAIL_USE_TLS") or True  # True
-MAIL_USE_SSL = os.getenv("MAIL_USE_SSL") or False  # False
+MAIL_PORT = int(os.getenv("MAIL_PORT")) if os.getenv("MAIL_PORT") else None  # 587
+MAIL_USE_TLS = os.getenv("MAIL_USE_TLS", "True") == "True" or True  # True
+MAIL_USE_SSL = os.getenv("MAIL_USE_SSL", "True") == "True" or False  # False
 MAIL_USERNAME = os.getenv("MAIL_USERNAME") or None  # some.email@test.se
 MAIL_PASSWORD = os.getenv("MAIL_PASSWORD") or None  # email_password
 
@@ -55,4 +55,4 @@ ADMINS = os.getenv("ADMINS") or ["test.email@mail.se"]
 
 # Set NOTIFY_COMPLETE to False if you don't want to notify variants and phenotypes by email
 # This way only contact info and matching patients ID will be notified in email body
-NOTIFY_COMPLETE = os.getenv("NOTIFY_COMPLETE") or False
+NOTIFY_COMPLETE = os.getenv("NOTIFY_COMPLETE", "False") == "True" or False

--- a/patientMatcher/utils/notify.py
+++ b/patientMatcher/utils/notify.py
@@ -123,6 +123,7 @@ def notify_match_internal(database, match_obj, admin_email, mail, notify_complet
         # send email using flask_mail
         try:
             mail.send(message)
+            LOG.info("Email notification sent")
         except Exception as err:
             LOG.error(
                 "An error occurred while sending an internal match notification: {}".format(err)
@@ -161,6 +162,7 @@ def notify_match_external(match_obj, admin_email, mail, notify_complete):
     # send email using flask_mail
     try:
         mail.send(message)
+        LOG.info("Email notification sent")
     except Exception as err:
         LOG.error("An error occurred while sending an external match notification: {}".format(err))
 

--- a/tests/server/test_create_app.py
+++ b/tests/server/test_create_app.py
@@ -7,6 +7,33 @@ from patientMatcher.server.__init__ import available_phenotype_resources
 from patientMatcher.utils.notify import TlsSMTPHandler
 
 
+def test_env_vars_parsing(monkeypatch):
+    """Test that the env vars are correctly parsed when creating the app"""
+    # GIVEN some params provided as env vars
+    monkeypatch.setenv("DEBUG", "True")
+    monkeypatch.setenv("TESTING", "False")
+    monkeypatch.setenv("DB_PORT", "27017")
+    monkeypatch.setenv("MAX_GT_SCORE", "0.7")
+    monkeypatch.setenv("MAX_PHENO_SCORE", "0.3")
+    monkeypatch.setenv("MAX_RESULTS", "3")
+    monkeypatch.setenv("SCORE_THRESHOLD", "0.5")
+    monkeypatch.setenv("MAIL_PORT", "586")
+    monkeypatch.setenv("NOTIFY_COMPLETE", "False")
+
+    app = create_app()
+
+    # THEN the app should be created with the expected config values
+    app.config["DEBUG"] is True
+    app.config["TESTING"] is False
+    app.config["DB_PORT"] == 27017
+    app.config["MAX_GT_SCORE"] == 0.7
+    app.config["MAX_PHENO_SCORE"] == 0.3
+    app.config["MAX_RESULTS"] == 3
+    app.config["SCORE_THRESHOLD"] == 0.5
+    app.config["MAIL_PORT"] == 586
+    app.config["NOTIFY_COMPLETE"] is False
+
+
 def test_create_app():
     """Tests the function that creates the app"""
 
@@ -32,29 +59,29 @@ def test_available_phenotype_resources_missing_resource():
     os.rename(temp_file, path_to_hpo_terms)
 
 
-def test_error_log_email(mock_app):
+def test_error_log_email(monkeypatch):
     """Test the app error logging via email"""
 
     # GIVEN an app with an ADMIN and configured email error logging params
-    mail_host = "smtp.gmail.com"
-    mail_port = 587
-    server_email = "server_email"
-    server_pw = "server_pw"
+    monkeypatch.setenv("ADMINS", '["app_admin_email"]')
+    monkeypatch.setenv("MAIL_SERVER", "smtp.gmail.com")
+    monkeypatch.setenv("MAIL_PORT", "587")
+    monkeypatch.setenv("MAIL_USERNAME", "server_email")
+    monkeypatch.setenv("MAIL_PASSWORD", "server_pw")
+    monkeypatch.setenv("MAIL_USE_TLS", "True")
 
-    mock_app.config["ADMINS"] = ["app_admin_email"]
-    mock_app.config["MAIL_SERVER"] = mail_host
-    mock_app.config["MAIL_PORT"] = mail_port
-    mock_app.config["MAIL_USERNAME"] = server_email
-    mock_app.config["MAIL_PASSWORD"] = server_pw
+    app = create_app()
 
-    configure_email_error_logging(mock_app)
+    configure_email_error_logging(app)
 
     # Then a TlsSMTPHandler should be among the app loggers
-    handler = mock_app.logger.handlers[0]
+    handler = app.logger.handlers[0]
     assert isinstance(handler, TlsSMTPHandler)
+
     # And should contain the given settings
-    assert handler.mailhost == mail_host
-    assert handler.mailport == mail_port
-    assert handler.fromaddr == server_email
-    assert handler.password == server_pw
-    assert handler.toaddrs == mock_app.config["ADMINS"]
+    assert handler.mailhost == "smtp.gmail.com"
+    assert handler.mailport == 587
+    assert handler.fromaddr == "server_email"
+    assert handler.password == "server_pw"
+    assert app.config["ADMINS"] == ["app_admin_email"]
+    assert isinstance(handler.toaddrs, list)

--- a/tests/server/test_create_app.py
+++ b/tests/server/test_create_app.py
@@ -3,6 +3,7 @@ import os
 
 from patientMatcher.resources import path_to_hpo_terms
 from patientMatcher.server import configure_email_error_logging, create_app
+from patientMatcher.server.__init__ import available_phenotype_resources
 from patientMatcher.utils.notify import TlsSMTPHandler
 
 
@@ -12,20 +13,20 @@ def test_create_app():
     assert create_app()
 
 
-def test_create_app_missing_requirement(monkeypatch):
-    """Tests the function that creates the app when phenotype_annotation.tab.txt and hp.obo.txt
-    resources are missing"""
+def test_available_phenotype_resources():
+    """Test function response when resource files (HPO terms and phenotype annotations) are both present"""
+    assert available_phenotype_resources()
 
-    # GIVEN a prod server
-    monkeypatch.setenv("TESTING", False)
 
-    # GIVEN one of the required files missing / it's been renamed
+def test_available_phenotype_resources_missing_resource():
+    """Test function response when one resource file is missing"""
+
+    # GIVEN one of the required files missing (here it's been renamed)
     temp_file = ".".join([path_to_hpo_terms, "temp"])
     os.rename(path_to_hpo_terms, temp_file)
 
-    # THEN the app should not be created
-    app = create_app()
-    assert app is None
+    # The function should return False
+    assert available_phenotype_resources() is False
 
     # Revert original resource name
     os.rename(temp_file, path_to_hpo_terms)

